mllib/src/test/scala/org/apache/spark/ml/regression/LinearRegressionSuite.scala

```diff
@@ -93,525 +96,641 @@ class LinearRegressionSuite extends SparkFunSuite with MLlibTestSparkContext {
-          Iterator(
-            Instance(label, weight = 1.2, features),
-            Instance(label, weight = 0.8, features)
-          )
-        case (false, LabeledPoint(label, features)) =>
-          Iterator(
-            Instance(label, weight = 0.3, features),
-            Instance(label, weight = 0.1, features),
-            Instance(label, weight = 0.6, features)
-          )
-      }
-
-      val noiseData = LinearDataGenerator.generateLinearInput(
-        2, Array(1, 3), Array(0.9, -1.3), Array(0.7, 1.2), 500, 1, 0.1)
-      val weightedNoiseData = noiseData.map {
-        case LabeledPoint(label, features) => Instance(label, weight = 0, features)
+    breakable {
+      Seq("auto", "l-bfgs", "normal").foreach { solver =>
+        val (data, weightedData) = {
+          val activeData = LinearDataGenerator.generateLinearInput(
+            6.3, Array(4.7, 7.2), Array(0.9, -1.3), Array(0.7, 1.2), 500, 1, 0.1)
+
+          val rnd = new Random(8392)
+          val signedData = activeData.map { case p: LabeledPoint =>
+            (rnd.nextGaussian() > 0.0, p)
+          }
+
+          val data1 = signedData.flatMap {
+            case (true, p) => Iterator(p, p)
+            case (false, p) => Iterator(p)
+          }
+
+          val weightedSignedData = signedData.flatMap {
+            case (true, LabeledPoint(label, features)) =>
+              Iterator(
+                Instance(label, weight = 1.2, features),
+                Instance(label, weight = 0.8, features)
+              )
+            case (false, LabeledPoint(label, features)) =>
+              Iterator(
+                Instance(label, weight = 0.3, features),
+                Instance(label, weight = 0.1, features),
+                Instance(label, weight = 0.6, features)
+              )
+          }
+
+          val noiseData = LinearDataGenerator.generateLinearInput(
+            2, Array(1, 3), Array(0.9, -1.3), Array(0.7, 1.2), 500, 1, 0.1)
+          val weightedNoiseData = noiseData.map {
+            case LabeledPoint(label, features) => Instance(label, weight = 0, features)
+          }
+          val data2 = weightedSignedData ++ weightedNoiseData
+
+          (sqlContext.createDataFrame(sc.parallelize(data1, 4)),
+            sqlContext.createDataFrame(sc.parallelize(data2, 4)))
+        }
+
+        val trainer1a = (new LinearRegression).setFitIntercept(true)
+          .setElasticNetParam(0.38).setRegParam(0.21).setStandardization(true).setSolver(solver)
+        val trainer1b = (new LinearRegression).setFitIntercept(true).setWeightCol("weight")
+          .setElasticNetParam(0.38).setRegParam(0.21).setStandardization(true).setSolver(solver)
+
+        var model1a0: LinearRegressionModel = null
+        var model1a1: LinearRegressionModel = null
+        var model1b: LinearRegressionModel = null
+
+        // Normal optimizer is not supported with non-zero elasticnet parameter.
+        if (solver == "normal") {
```

Comment:
Yes, either change it to 0.0 or have a copy of it. I think change it to 0.0 should be fine.